arch/riscv/include/kernel_arch_func.h

```diff
@@ -45,7 +45,15 @@ FUNC_NORETURN void z_riscv_fatal_error(unsigned int reason,
 
 static inline bool arch_is_in_isr(void)
 {
+#ifdef CONFIG_SMP
+	unsigned int key = arch_irq_lock();
```

Comment:
You can actually skip this lock.  If you are in an ISR, then by definition you won't migrate to another CPU, and thus the access to the current cpu struct is race-free.  If you are not in an ISR, then the value of nested is by definition zero.

But yes, all access to the cpu struct must be atomic.  In the kernel we typically use _current_cpu, which is a macro that  includes an assertion checking this condition.

Reply:
But isn't this function... actually verifying if we're in an ISR? What if we aren't?

Reply:
Never mind, I was wrong. :)  The thinking was: that if you were in the ISR you couldn't be preempted, so it would always report true regardless of locking.   And if you weren't, it would always report false by definition.  So thus no lock needed.

But the lock isn't to protect the result, it's to protect the CPU pointer.  Without the lock, you can check the current CPU, and get an answer, and then get preempted and migrated to another CPU.  And so when you pull out the nested value to see if you're in an ISR you'll get the value of is_in_isr() of **some other CPU on which you are not running**. 

So you were right and I was wrong.  Leave the lock :)